core/chains/evm/gas/arbitrum_estimator.go

```diff
@@ -0,0 +1,168 @@
+	"golang.org/x/exp/slices"
+
+	evmclient "github.com/smartcontractkit/chainlink/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+type ethClient interface {
+	CallContract(ctx context.Context, msg ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+}
+
+// arbitrumEstimator is an Estimator which extends l2SuggestedPriceEstimator to use getPricesInArbGas() for gas limit estimation.
+type arbitrumEstimator struct {
+	utils.StartStopOnce
+
+	Estimator // *l2SuggestedPriceEstimator
+
+	client     ethClient
+	pollPeriod time.Duration
+	logger     logger.Logger
+
+	getPricesInArbGasMu  sync.RWMutex
+	perL2Tx              uint32
+	perL1CalldataUnit    uint32
+	perStorageAllocation uint32 //TODO unused?
+
+	chForceRefetch chan (chan struct{})
+	chInitialised  chan struct{}
+	chStop         chan struct{}
+	chDone         chan struct{}
+}
+
+// TODO config for max gas limit?
+func NewArbitrumEstimator(lggr logger.Logger, rpcClient rpcClient, ethClient ethClient) Estimator {
+	lggr = lggr.Named("ArbitrumEstimator")
+	return &arbitrumEstimator{
+		Estimator:      NewL2SuggestedPriceEstimator(lggr, rpcClient),
+		client:         ethClient,
+		pollPeriod:     10 * time.Second,
+		logger:         lggr,
+		chForceRefetch: make(chan (chan struct{})),
+		chInitialised:  make(chan struct{}),
+		chStop:         make(chan struct{}),
+		chDone:         make(chan struct{}),
+	}
+}
+
+func (a *arbitrumEstimator) Start(ctx context.Context) error {
+	return a.StartOnce("ArbitrumEstimator", func() error {
+		if err := a.Estimator.Start(ctx); err != nil {
+			return errors.Wrap(err, "failed to start gas price estimator")
+		}
+		go a.run()
+		<-a.chInitialised
+		return nil
+	})
+}
+func (a *arbitrumEstimator) Close() error {
+	return a.StopOnce("ArbitrumEstimator", func() (err error) {
+		close(a.chStop)
+		err = errors.Wrap(a.Estimator.Close(), "failed to stop gas price estimator")
+		<-a.chDone
+		return
+	})
+}
+
+// GetLegacyGas uses the gas price estimate
+func (a *arbitrumEstimator) GetLegacyGas(calldata []byte, l2GasLimit uint32, maxGasPriceWei *big.Int, opts ...Opt) (gasPrice *big.Int, chainSpecificGasLimit uint32, err error) {
+	gasPrice, _, err = a.Estimator.GetLegacyGas(calldata, l2GasLimit, maxGasPriceWei, opts...)
+	if err != nil {
+		return
+	}
+	ok := a.IfStarted(func() {
+		if slices.Contains(opts, OptForceRefetch) {
+			ch := make(chan struct{})
+			a.chForceRefetch <- ch
+			select {
+			case <-ch:
+			case <-a.chStop:
+				err = errors.New("estimator stopped")
+				return
+			}
+		}
+		chainSpecificGasLimit = a.calculateGas(len(calldata), l2GasLimit)
+		a.logger.Debugw("GetLegacyGas", "calldataLen", len(calldata), "l2GasLimit", l2GasLimit, "chainSpecificGasLimit", chainSpecificGasLimit)
+	})
+	if !ok {
+		return nil, 0, errors.New("estimator is not started")
+	}
+	//TODO enforce a maximum? (limit, or overall fee? txm could limit overall fee instead....)
```

Comment:
We could do the same as the price estimator, return an error if over some configured max:
https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/gas/l2_suggested_estimator.go#L158-L160

Reply:
I would suggest doing this, configured max is supposed to be an inviolable hard upper limit.

Reply:
We don't have a max gas limit config var AFAIK. If we add one, should it just be global/per-chain, or does it also need to be overridable at the key-specific level too? The former is simpler to implement, by wiring in standard config interfaces, but IIUC the latter would require changing the method signature to pass in a max limit just like the `maxGasPriceWei`, and I'm not sure what is required upstream from there.